Convert util/{HttpHeaderLink,PairedList,TimestampUtils,ThemeUtils} to Kotlin

nikclayton opened the pull request.

connyduck commented on app/src/main/java/com/keylesspalace/tusky/util/ThemeUtils.kt:
```diff
@@ -0,0 +1,74 @@
+ * Provides runtime compatibility to obtain theme information and re-theme views, especially where
+ * the ability to do so is not supported in resource files.
+ */
+object ThemeUtils {
```
`ThemeUtils`/`TimeStampUtils` shouldn't be objects but top-level functions.

`object` in Kotlin is similar to a singleton pattern in Java, the class will be instantiated and kept around at runtime, which makes no sense if it does not hold any state.
Top-level functions on the other hand are like static functions in Java (=what they were before the conversion) and thus have less overhead. With an `@file:JvmName` annotation you can choose the name for the class that holds the static functions and if you set it to `ThemeUtils`/`TimeStampUtils` no Java callsites have to be updated. See e.g. `LinkHelper` for that pattern.

connyduck commented on app/src/main/java/com/keylesspalace/tusky/util/HttpHeaderLink.kt:
```diff
@@ -0,0 +1,150 @@
+
+    private val parameters: MutableList<Parameter> = ArrayList()
+
+    var uri: Uri
+
+    init {
+        this.uri = Uri.parse(uri)
+    }
```
```suggestion
    val uri: Uri = uri.toUri()
```

connyduck commented on app/src/main/java/com/keylesspalace/tusky/util/HttpHeaderLink.kt:
```diff
@@ -0,0 +1,150 @@
+            links: List<HttpHeaderLink>,
+            relationType: String
+        ): HttpHeaderLink? {
+            for (link in links) {
+                for (parameter in link.parameters) {
+                    if (parameter.name == "rel" && parameter.value == relationType) {
+                        return link
+                    }
+                }
+            }
+            return null
```
You can write this in nice idiomatic Kotlin like so:

```suggestion
         return links.find { link ->
                link.parameters.any { parameter ->
                    parameter.name == "rel" && parameter.value == relationType
                }
            }
```

connyduck commented on app/src/main/java/com/keylesspalace/tusky/util/HttpHeaderLink.kt:
```diff
@@ -0,0 +1,150 @@
+         * @return all links found in the header
+         */
+        fun parse(line: String?): List<HttpHeaderLink> {
+            val linkList: MutableList<HttpHeaderLink> = ArrayList()
```
if it doesn't have to be an `ArrayList` better use `mutableListOf()`

connyduck commented on app/src/main/java/com/keylesspalace/tusky/util/HttpHeaderLink.kt:
```diff
@@ -0,0 +1,150 @@
+ * @see [RFC5988](https://tools.ietf.org/html/rfc5988)
+ */
+class HttpHeaderLink private constructor(uri: String) {
+    data class Parameter(var name: String? = null, var value: String? = null)
```
I am sure you can make these both `val` if you restructure the code a bit

connyduck commented on app/src/main/java/com/keylesspalace/tusky/util/HttpHeaderLink.kt:
```diff
@@ -0,0 +1,150 @@
+ * neighboring rights to this software to the public domain worldwide. This software is distributed
+ * without any warranty.
+ *
+ * You should have received a copy of the CC0 Public Domain Dedication along with this software.
```
uhm that is not our license

nikclayton replied:
It's the license on the most recent version of the code, https://github.com/tuskyapp/Tusky/blob/7a2ee150788ed1e1353214f125c418a9eae6969e/app/src/main/java/com/keylesspalace/tusky/util/HttpHeaderLink.java.

I can change it if you think that's appropriate.

connyduck replied:
Probably an artifact of the time when Vavassor tried to do a license switch, or maybe he copied the code from somewhere else. 
Please change it to 2022 Tusky contributors and GPL

connyduck commented on app/src/main/java/com/keylesspalace/tusky/util/HttpHeaderLink.kt:
```diff
@@ -0,0 +1,150 @@
+        }
+
+        private fun parseValue(line: String, start: Int): ValueResult {
+            val result = ValueResult()
+            val foundIndex = line.indexOfAny(charArrayOf(';', ',', '"'), start, false)
+            if (foundIndex == -1) {
+                result.value = line.substring(start)
+                return result
+            }
```
How about making `value` a constructor param of `ValueResult` and then something like

```suggestion
            val foundIndex = line.indexOfAny(charArrayOf(';', ',', '"'), start, false)
            if (foundIndex == -1) {
                return ValueResult(line.substring(start))
            }
```

connyduck commented on app/src/test/java/com/keylesspalace/tusky/util/HttpHeaderLinkTest.kt:
```diff
@@ -0,0 +1,83 @@
+    @Test
+    fun shouldParseValidLinks() {
+        // Constructor is private, so call through reflection
+        val constructor = HttpHeaderLink::class.primaryConstructor!!
```
Usually we just make public what we need and annotate it with `@VisibleForTesting`

connyduck commented on app/src/main/java/com/keylesspalace/tusky/util/PairedList.kt:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    fun getPairedItemOrNull(index: Int): V? {
+        return if (index >= 0 && index < synced.size) {
+            synced[index]
+        } else {
+            null
+        }
```
```suggestion
        return synced.getOrNull(index)
```

nikclayton commented:
Thanks for the helpful feedback. Changes made.

I removed `ThemeUtils.getColors` as well, and replaced that with the one that's already in the `MaterialColors` library.
